### DISCUSSION TITLE
端到端传输链路上的时延

### DISCUSSION THREAD

**Discussion by stopandjoin** [Tencent/tquic#294] [Q&A (问题和回答)] [answered]
请教一下，默认情况下，tquic stream的发送端和接收端的内部有多大的buffer，最多会产生多大的时延？  
我在自己的项目发现一种情况，两条stream A 和 B，优先级A高于B， A和B同时发送数据， 保证发送端都正常写入了（quic_stream_write 返回正确）。弱网情况下， 在接收端，stream A收到数据没什么明显的时延， 但是stream B，接收到数据时，时延很大（可以达到10秒以上），可能是哪里缓冲住了这么久吗？

**Comment by glime-ai** [comment]
在TQUIC中，默认情况下，发送端和接收端的内部缓冲区大小以及可能产生的时延会受到多种因素的影响，包括网络状况、流量控制机制、拥塞控制算法等。

1. **缓冲区大小**：TQUIC的发送端和接收端的缓冲区大小并没有固定的默认值，而是可以根据具体的实现和配置进行调整。一般来说，发送端会有一个发送缓冲区，用于存储待发送的数据；接收端会有一个接收缓冲区，用于存储已接收但尚未处理的数据。

2. **时延问题**：在弱网环境下，时延可能会显著增加。根据你的描述，stream A和B同时发送数据，但stream B的时延明显大于stream A，这可能是由于以下几个原因：
   - **优先级设置**：虽然stream A的优先级高于stream B，但在弱网环境下，网络抖动、丢包等问题可能会导致低优先级的stream B的数据被延迟处理。
   - **流量控制**：TQUIC使用流量控制机制来管理每个流的发送和接收速率。如果stream B的数据量较大，可能会受到流量控制窗口的限制，从而导致时延增加。
   - **拥塞控制**：TQUIC采用拥塞控制算法（如CUBIC、BBR等）来调节发送速率。在网络状况不佳时，拥塞控制算法可能会降低发送速率，从而增加时延。

为了更好地理解和解决这个问题，建议你检查以下几点：
- **缓冲区配置**：查看并调整发送端和接收端的缓冲区大小配置，确保其适合当前的网络环境。
- **流量控制窗口**：检查流量控制窗口的设置，确保其不会过于限制数据的发送和接收。
- **拥塞控制算法**：了解并选择适合当前网络状况的拥塞控制算法，以优化数据传输性能。

你可以参考以下文档获取更多信息：
- [TQUIC 性能测试](https://tquic.net/zh/docs/further_readings/benchmark)
- [TQUIC 类似项目比较](https://tquic.net/zh/docs/further_readings/comparison)

This comment was generated by [Glime](https://glimelab.ai/?utm_source=44520fce49ba43798485a628d6928e22&utm_medium=github).

**Comment by iyangsj** [comment]
@stopandjoin 发送端只要可发送，就会发送当前已缓冲端数据。把完整的debug级别日志发出来看看。
看起来发送端由于受到流量控制/拥塞控制限制，导致无法发送（例如弱网环境下流控窗口更新帧，或确认帧持续出现丢包重传）。

**Reply by stopandjoin** [reply]
好的，我复现一下抓个日志。

**Reply by stopandjoin** [reply]
@iyangsj  复现概率非常低，发现是在网络带宽非常低的时候出现的，应该就是你推测的“确认帧持续出现丢包重传“”。 
这种情况下，有什么优化建议吗？ 比如调节发包的大小是否能有改善。

**Reply by iyangsj** [reply]
可以添加微信`iyangsj`讨论

**Comment by stopandjoin** [comment]
[tquic_client.txt](https://github.com/user-attachments/files/16455249/tquic_client.txt)

**Comment by stopandjoin** [comment]
[tquic_server.txt](https://github.com/user-attachments/files/16455415/tquic_server.txt)

**Comment by stopandjoin** [comment]
@iyangsj 复现了一次，以上是两端的tquic日志， 时间点在18:21 前后。

**Comment by stopandjoin** [comment]
找到了复现路径： 将带宽限制到很低，1Mbps以下然后连接。 刚连接上前2分钟内，很容易复现。   可能是写到tquic 内部buffer中的数据在慢慢发送。 我们现在是实时音视频场景，看tquic的api，好像没有主动清理掉 tquic 内部的buffer的接口。 是否还有其他办法。
刚有复现到一次，log如：
[0802_server.log](https://github.com/user-attachments/files/16464330/0802_server.log)
[0802_client.log](https://github.com/user-attachments/files/16464333/0802_client.log)

**Reply by iyangsj** [reply]
> 好像没有主动清理掉 tquic 内部的buffer的接口。 是否还有其他办法

无法清除已写入流缓冲区的数据（这部分数据可能已发送待确认，清理掉并不符合协议规范）。只能关闭流，建立新的流。

> 两条stream A 和 B，优先级A高于B

采用相同优先级，且使用incremental模式，能尽可能保证两条流数据以同等优先级发送，最小化接收端的时间差

**Reply by stopandjoin** [reply]
了解啦。不过业务需求是A优先级要最高。 较低优先级的数据，时延是没法保证吧，那有没有接口，可以把这条流的buffer设置小一点？这样可以在业务层主动丢包。 @iyangsj

**Reply by stopandjoin** [reply]
@iyangsj  或者能不能只对延时大的这条stream 执行RESET_STREAM ，然后重新建立？

**Reply by iyangsj** [reply]
使用[shutdown_stream](https://docs.rs/tquic/latest/tquic/connection/struct.Connection.html#method.stream_shutdown) 可以强制Reset关闭流

**Reply by iyangsj** [reply, marked as answer]
> 了解啦。不过业务需求是A优先级要最高。 较低优先级的数据，时延是没法保证吧，那有没有接口，可以把这条流的buffer设置小一点？这样可以在业务层主动丢包。 

后续版本会增加相关接口

**Reply by stopandjoin** [reply]
> 使用[shutdown_stream](https://docs.rs/tquic/latest/tquic/connection/struct.Connection.html#method.stream_shutdown) 可以强制Reset关闭流
我尝试过了，shutdown stream后， 使用相同 id 重新 new stream， 但是不成功。代码中这个逻辑会影响吧：
![image](https://github.com/user-attachments/assets/0381e9a9-6d86-4c8b-b20c-7cbbaae7cd9a)

> 后续版本会增加相关接口
如果我们自己修改tquic代码，新增这个接口，需要改哪个地方，是不是设置一下这个max_data：
![lQLPKG4gvAGaBc_NAVDNAyWwGy6mcaXbDpEGndYCZWNrAA_805_336](https://github.com/user-attachments/assets/b8049a58-2791-4eb2-9db3-6009c321c6ed)

@iyangsj

**Reply by stopandjoin** [reply]
新发现一个问题：在client端关闭stream调用 quic_stream_shutdown 会偶尔失效，表现情况有两种：
1，server端可以收到 stream closed 回调， 但是 client端没有  stream closed 回调。stream实际断了。
2， 两端都没有 stream closed 回调。stream依旧保持着。

client端调用方法：
        int ret_w = quic_stream_shutdown(mConn,id,QUIC_SHUTDOWN_WRITE,0);
        int ret_r = quic_stream_shutdown(mConn,id,QUIC_SHUTDOWN_READ,0);

**Reply by stopandjoin** [reply]
@iyangsj  补充一下信息，tquic版本是0.8.0， client和server两端都是Android平台，NDK 23
